src/redisraft.c

```diff
@@ -591,76 +591,80 @@ static bool handleInterceptedCommands(RedisRaftCtx *rr,
-             * Otherwise, we might process a request before replaying logs.
-             * The state machine will be in an older state. Reading from it
-             * might look like going backward in time. */
-            raft_term_t term = raft_get_current_term(rr->raft);
-            if (term != rr->snapshot_info.last_applied_term) {
-                replyClusterDown(ctx);
-                return RR_ERROR;
-            }
-
+    if (rr->config->quorum_reads) {
+        RaftReq *req = RaftReqInit(ctx, RR_REDISCOMMAND);
+        RaftRedisCommandArrayMove(&req->r.redis.cmds, cmds);
+        raft_queue_read_request(rr->raft, handleReadOnlyCommand, req);
+    } else {
+        /* Wait until the new leader applies an entry from the current term.
+         * Otherwise, we might process a request before replaying logs.
+         * The state machine will be in an older state. Reading from it
+         * might look like going backward in time. */
+        raft_term_t term = raft_get_current_term(rr->raft);
+        if (term == rr->snapshot_info.last_applied_term) {
             RaftExecuteCommandArray(rr, ctx, ctx, cmds);
+        } else {
+            replyClusterDown(ctx);
         }
-        return RR_OK;
     }
+}
 
+static void handleRedisCommandAppendLog(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommandArray *cmds) {
     RaftReq *req = RaftReqInit(ctx, RR_REDISCOMMAND);
     RaftRedisCommandArrayMove(&req->r.redis.cmds, cmds);
 
     raft_entry_t *entry = RaftRedisCommandArraySerialize(&req->r.redis.cmds);
     entry->id = rand();
     entry->type = RAFT_LOGTYPE_NORMAL;
     entryAttachRaftReq(rr, entry, req);
+
     int e = raft_recv_entry(rr->raft, entry, &req->r.redis.response);
     if (e != 0) {
         replyRaftError(ctx, e);
         entryDetachRaftReq(rr, entry);
         raft_entry_release(entry);
         RaftReqFree(req);
-        return RR_ERROR;
+    } else {
+        raft_entry_release(entry);
     }
+}
 
-    raft_entry_release(entry);
-    return RR_OK;
+static void handleRedisCommandAppend(RedisRaftCtx *rr,
+                                      RedisModuleCtx *ctx, RaftRedisCommandArray *cmds)
+{
+    unsigned int cmd_flags = CommandSpecGetAggregateFlags(cmds, CMD_SPEC_WRITE);
+
+    /* Check that we're part of a bootstrapped cluster and not in the middle of
+     * joining or loading data.
+     */
+    if (checkRaftState(rr, ctx) == RR_ERROR) {
+        return;
+    }
+
+    if (!raft_is_leader(rr->raft)) {
+        Node *leader = getLeaderNodeOrReply(rr, ctx);
+
+        return handleNonLeaderCommand(rr, ctx, leader, cmds);
+    }
+
+    if (cmd_flags & CMD_SPEC_UNSUPPORTED) {
+        RedisModule_ReplyWithError(ctx, "ERR not supported by RedisRaft");
+    } else if (cmd_flags & CMD_SPEC_READONLY && !(cmd_flags & CMD_SPEC_WRITE)) {
+        handleRedisCommandAppendReadOnly(rr, ctx, cmds);
+    } else {
+        handleRedisCommandAppendLog(rr, ctx, cmds);
+    }
```

Comment:
```suggestion
    if (cmd_flags & CMD_SPEC_UNSUPPORTED) {
        RedisModule_ReplyWithError(ctx, "ERR not supported by RedisRaft");
        return;
    } 
    
    if (cmd_flags & CMD_SPEC_READONLY && !(cmd_flags & CMD_SPEC_WRITE)) {
        handleRedisCommandAppendReadOnly(rr, ctx, cmds);
    } else {
        handleRedisCommandAppendLog(rr, ctx, cmds);
    }
```

For example, I find this one easier to understand.

Reply:
ok, done